drivers/samsung/displays/md_series.cr

```diff
@@ -0,0 +1,500 @@
+    Hdmi3         = 0x31
+    Hdmi3_pc      = 0x32
+    Hdmi4         = 0x33
+    Hdmi4_pc      = 0x34
+    Display_port  = 0x25
+    Dtv           = 0x40
+    Media         = 0x60
+    Widi          = 0x61
+    Magic_info    = 0x20
+    Whiteboard    = 0x64
+  end
+
+  def switch_to(input : String, **options)
+    self[:input_stable] = false
+    self[:input_target] = input
+    do_send("input", INPUTS.parse(input).value, **options)
+  end
+
+  enum SCALEMODE
+    Fill = 0x09
+    Fit =  0x20
+  end
+
+  # Activite the internal compositor. Can either split 3 or 4 ways.
+  def split(inputs : Array(String) = ["hdmi", "hdmi2", "hdmi3"], layout : Int32 = 0, scale : String = "fit", **options)
+    main_source = inputs.shift
+
+    data = [
+      1,                  # enable
+      0,                  # sound from screen section 1
+      layout,             # layout mode (1..6)
+      SCALEMODE.parse(scale).value,  # scaling for main source
+      inputs.flat_map do |input|
+        [INPUTS.parse(input).value, SCALEMODE.parse(scale).value]
+      end
+    ].flatten
+
+    switch_to(main_source, **options)
+    do_send("screen_split", data, **options)
+  end
+
+  def in_range(val : Int32, max : Int32) : Int32
+    min = 0
+    if val < min
+      val = min
+    elsif val > max
+      val = max
+    end
+    val
+  end
+
+  # Emulate mute
+  def mute_audio(val : Bool = true)
+    if val
+      if !self[:audio_mute]?
+        self[:audio_mute] = true
+        self[:previous_volume] = self[:volume].as_i? || 50
+        volume(0)
+      end
+    else
+      unmute_audio
+    end
+  end
+
+  def unmute_audio
+    if self[:audio_mute]?
+      self[:audio_mute] = false
+      volume(self[:previous_volume].as_i? || 50)
+    end
+  end
+
+  enum SPEAKERMODES
+    Internal = 0
+    External = 1
+  end
+
+  def speaker_select(mode : String, **options)
+    do_send("speaker", SPEAKERMODES.parse(mode).value, **options)
+  end
+
+  def do_poll
+    do_send("status", [] of UInt8, priority: 0)
+    power? unless self[:hard_off]?
+  end
+
+  # Enable power on (without WOL)
+  def network_standby(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("net_standby", state, **options)
+  end
+
+  # Eco auto power off timer
+  def auto_off_timer(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("eco_solution", [0x81, state], **options)
+  end
+
+  # Device auto power control (presumably signal based?)
+  def auto_power(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send("auto_power", state, **options)
+  end
+
+  # Display control
+  METHODS = [
+    "volume",
+    "contrast",
+    "brightness",
+    "sharpness",
+    "colour",
+    "tint",
+    "red_gain",
+    "green_gain",
+    "blue_gain"
+  ]
+
+  # Macro to define methods from the array above
+  {% for name in METHODS %}
+    def {{name.id}}(val : Int32, **options)
+      val = in_range(val, 100)
+      do_send({{name.id.stringify}}, val, **options)
+    end
+  {% end %}
```

Comment:
When using macros like this, unless it's needed elsewhere, you can just keep the collection that you're iterating over directly in the macro to keep the resulting code clean.

Reply:
Hey @KimBurgess , I'm not sure how to make this any cleaner. I'm using `METHODS` in the first line of my macro so not sure where I would define the collection elsewhere.